### DISCUSSION TITLE
Why does SvelteKit not generate the types for $page.data?

### DISCUSSION THREAD

**Discussion by dan-cooke** [sveltejs/kit#10002] [Q&A]
I'm just wondering why all the routes have their own `PageData` defined which is accurate, but the root `App.PageData` cannot be generated by combining these types?

Its most likely intentional, but I'd love to know more about the reasons for the design choice

**Comment by david-plugge** [comment]
Id would be rather hard to get the types 100% correct.

```ts
type PageOne = {
	a: number;
	b: number;
	c: string;
}

type PageTwo = {
	b: number;
	c: number;
	d: string
}
```
The naive approach would be to `|` all types together like so:

```ts
type PageData = PageOne | PageTwo
```

Copy the code into your editor and see that this is not working as expected:
![image](https://github.com/sveltejs/kit/assets/59972093/436bcfc6-aeef-4c89-8279-e7dcf65492f4)

Only properties `b` and `c` are shown as they appear on both types.

To get around this we would need to add the missing properties to ALL types:

```ts
type PageOne = {
	a: number;
	b: number;
	c: string;

	d?: undefined;
}

type PageTwo = {
	b: number;
	c: number;
	d: string;

	a?: undefined;
}
```

This works!
![image](https://github.com/sveltejs/kit/assets/59972093/d95c491c-f4b0-4cd5-86f7-4442444e0bc2)

But this is a very simple example. You can imagine how complex this can become with more pages and nested data.
You _can_ achieve this behaviour with some advanced typescript stuff, but it´s probably not worth it and i dont think the typescript server would like the extra work at all.

`export let data: PageData` is typed, just use that whenever you can and if not, just add types by yourself.

I often end up splitting my `+page.svelte` file into components and use the `PageData`:

_/posts/+page.server.ts_
```ts
export const load = () => {
	return {
		posts: db.getPostsSomehow()
	}
}
```

_/posts/Post.svelte_
```svelte
<script lang="ts">
	import type { PageData } from './$types';

	export let post: PageData['posts'][number];
</script>
...
```

_/posts/+page.svelte_
```svelte
<script lang="ts">
	import Post from './Post.svelte';

	export let data;
	import type { PageData } from './$types';
</script>

{#each data.posts as post}
	<Post {post} />
{/each}
```

The same goes for data coming from layout loads. Just import the correct `LayoutData` and you are all set.
